### PR TITLE
[mem_cache] feat: reduce SLRU burst distortion and stagnation

### PR DESCRIPTION
**Pull request by HughLLiu**
# Improve SLRU radix-cache eviction with debounce and lazy decay

## Motivation

The current SLRU radix-cache eviction policy promotes a prefix once its `hit_count` reaches the Protected threshold. This works well for stable hot prefixes, but it has two failure modes under shared-prefix and high-pressure KV-cache workloads:

1. **Burst distortion**: many concurrent requests can hit the same short-lived prefix within a few milliseconds, inflating `hit_count` and promoting data that may never be reused.
2. **Zombie Protected entries**: once a prefix enters Protected, historical heat never decays, so stale prefixes can stay protected long after the workload has shifted.

This PR does not try to make SLRU universally better than LRU. It makes SLRU less fragile in workloads where SLRU is already the intended policy, especially workloads with reusable shared prefixes.

## Modifications

This PR keeps the existing SLRU policy shape, but changes hit accounting when `SGLANG_ENABLE_SLRU_OPTIMIZATION=1`:

- Add write-side debounce: hits inside `slru_debounce_sec` do not increment `hit_count`.
- Cap protected nodes at `slru_protected_threshold`: Protected nodes refresh their timestamp but do not accumulate unbounded historical heat.
- Add eviction-time lazy decay: SLRU computes an effective hit count during eviction.

```text
effective_hit_count = hit_count >> floor(age / slru_decay_sec)
```

If the effective hit count falls below `slru_protected_threshold`, the node is treated as Probationary for that eviction scan.

The default knobs are:

- `slru_protected_threshold = 2`
- `slru_debounce_sec = 0.1`
- `slru_decay_sec = 60.0`

For GSP-style shared-prefix workloads, a shorter debounce window such as `0.01s` or `0.02s` can be better because true hot prefixes need to enter Protected quickly. For more heterogeneous trace workloads such as Mooncake, `0.1s` is often a better default because many short-interval hits are burst artifacts rather than durable prefix reuse.

## Accuracy Tests

Not applicable. This PR changes radix-cache eviction priority and hit accounting. It does not change model weights, kernels, sampling, logits, or model forward computation.

### GSP shared-prefix benchmark

Config:

```text
Model: Qwen/Qwen2.5-7B-Instruct
--gsp-num-groups=120
--gsp-prompts-per-group=70
--gsp-system-prompt-len=2048
--request-rate=3
SLRU knobs: threshold=2, debounce=0.02s, decay=60.0s
```

On a saturated GSP shared-prefix workload, optimized SLRU reduced p99 TTFT relative to naive SLRU:

| policy | p99 TTFT |
| --- | ---: |
| lru | 13869.53 ms |
| naive-slru | 19160.66 ms |
| optimized-slru | 15418.49 ms |

Optimized SLRU improved p99 TTFT vs naive SLRU by:

```text
(15418.49 - 19160.66) / 19160.66 = -19.53%
```

This recovers a large part of naive SLRU's tail regression under bursty shared-prefix pressure, although LRU is still the best policy in this particular run.

On the same type of GSP shared-prefix workload, optimized SLRU also improved median TTFT and prefix-cache reuse:

| metric | naive SLRU | optimized SLRU | delta |
| --- | ---: | ---: | ---: |
| median TTFT | 533.88 ms | 485.76 ms | -9.01% |
| prefix-cache hit ratio | 0.2911 | 0.3009 | +0.0098 |
| cached tokens | 5,034,434 | 5,196,267 | +161,833 |
| new prefill tokens | 12,260,897 | 12,073,571 | -187,326 |
| retracts | 44 | 44 | tied |

The equal retract count indicates comparable physical KV pressure. The latency improvement comes from better cache reuse rather than reduced load.

### Mooncake benchmark

In a larger Mooncake run with default SLRU knobs, optimized SLRU improved both median and p99 TTFT relative to naive SLRU:

| metric | naive SLRU | optimized SLRU | delta |
| --- | ---: | ---: | ---: |
| median TTFT | 183321.93 ms | 179673.90 ms | -2.0% |
| p99 TTFT | 358232.13 ms | 351048.60 ms | -2.0% |
| output tok/s | 463.62 | 464.81 | +0.3% |

In a low-pressure Mooncake run, LRU, naive SLRU, and optimized SLRU were within roughly 0.2%, suggesting the optimization does not introduce measurable overhead when the failure mode is absent.

### Limitations

This PR does not claim that SLRU is always better than LRU.

In fast-churn or tight-KV workloads, LRU can still outperform both naive and optimized SLRU. For example, in a BurstGPT-derived tight-KV run, optimized SLRU improved median TTFT relative to naive SLRU, but both SLRU variants were still slower than LRU.

The intended claim is narrower:

- naive SLRU has burst and stale-hot-prefix failure modes
- this PR reduces those failure modes
- the best eviction policy can still depend on workload shape

## Unit Tests

Added unit coverage for:

- legacy SLRU behavior with the optimization gate disabled
- debounce preventing burst prefixes from entering Protected
- hit-count cap behavior
- lazy decay demoting stale Protected nodes
- radix-tree split preserving the timestamp needed by lazy decay

Suggested local validation:

```bash
python3 -m pytest test/registered/unit/mem_cache/test_evict_policy.py -q
python3 -m pytest test/registered/unit/mem_cache/test_slru_hit_rate.py -q
python3 -m pytest test/registered/unit/mem_cache -q
pre-commit run --all-files
```

## Checklist

- [ ] Format code according to the SGLang pre-commit guidance.
- [ ] Add or update unit tests for radix-cache eviction behavior.
- [ ] Update documentation if maintainers want these SLRU knobs documented publicly.
- [ ] Provide speed benchmark results for shared-prefix and trace-style workloads.
- [ ] Follow SGLang code style guidance, especially keeping runtime-path overhead small.

## Review and Merge Notes

This change touches KV-cache code under `python/sglang/srt/mem_cache`, so the relevant reviewers are the KV Cache maintainers listed in `.github/MAINTAINER.md`.

After opening the PR:

1. Wait for the bot to assign a Merge Oncall and request Codeowner review.
2. If CI does not start, ask an authorized maintainer to add the `run-ci` label or run `/tag-and-rerun-ci`.
3. As the PR author, use `/rerun-failed-ci` if a failed check appears flaky after CI has run.

### CODE REVIEW

**Comment by HughLLiu** [comment]
Hi — could someone please add the run-ci label when you have a moment so CI can run? Thanks!
